### ISSUE TITLE
Toggling prerelease icon with caching enabled

### ISSUE THREAD

**Issue by icleolion** [vscode-contrib/vscode-versionlens#250]
Version lenses don't update when prerelease icon is toggled until cache is invalidated.

Rather than lose the benefit of the cache by making it so toggling the prerelease icon flushes the cache, would two caches, one for each mode help?

Or have every fetch of package info store the prerelease tags into cache anyway, and just show/hide that part of the info in the lens based on the icon state?

**Comment by icleolion** [comment]
oh and while i'm here, typo...
![image](https://user-images.githubusercontent.com/47498402/83281618-43252a00-a1d0-11ea-816d-39950fad1e10.png)

**Comment by pflannery** [comment]
> Version lenses don't update when prerelease icon is toggled until cache is invalidated.
> 
> Rather than lose the benefit of the cache by making it so toggling the prerelease icon flushes the cache, would two caches, one for each mode help?
> 
> Or have every fetch of package info store the prerelease tags into cache anyway, and just show/hide that part of the info in the lens based on the icon state?

Its a good idea.

The prereleases come with releases when each package is fetched so the icon is only showing or hiding them. There are a couple of possiblities I can think of:

- include the prerelase icon state as part of the key used to store the cached entry
- reduce the default cache duration to 1 minute. 
- cache per document session e.g cache resets next time the document is opened.
- have a shift+click action on icon to enforce a fresh request

I implemented the caching to:

- give a better perf
- reduce continous requests going to the registries (eg. each edit in the document refreshes the lenses and thus makes requests). RequestCount = EditCount*packageCount (if cache duration is 0)

I dont think people spend too much time checking versions but when they do they want the know the very latest state of their dependencies (at least I do). 

Its difficult to gauge the user need as there isn't any user research on this project being a front room project. If I had sponsorship I could invest in to user research.

For now I've been thinking about "opt-in" anonymous analytics and\or surveys to try and gauge what people need but that type of research is very black and white (difficult to read a person via textual input)

right now I'm thinking the shift+click option could be a winner.

**Comment by icleolion** [comment]
Changing cache to 1min is the first thing I did. Plus I have an issue with the way that works in that the cache isn’t flushed out until a change to document occurs. Turning Version Lens off and on and/or toggling the prerelease icon doesn’t force a new fetch. This issue is how I spotted the original issue in this first place.

I check my versions probably three times a day and yeah I expect the process to give me the latest versions without fuss so that it takes me not much longer than the time it takes to scroll through the package list to know if I’ve got any packages I need to seek out changelogs for.

As for the prerelease icon already just doing a show/hide, that’s not what i’m seeing in the output, as the number of dependencies resolved increases when prerelease icon is turned on.

**Comment by icleolion** [comment]
As for shift+click, considering these are icons that only appear for specific files, I don't the space is an issue, so could a third icon be added that only shows up when caching is enabled that on click flushes the cache, rather than hiding it away as a key combo?

**Comment by pflannery** [comment]
> As for shift+click, considering these are icons that only appear for specific files, I don't the space is an issue, so could a third icon be added that only shows up when caching is enabled that on click flushes the cache, rather than hiding it away as a key combo?

yeah it could be done that way.

**Comment by icleolion** [comment]
How do the different caching options relate to each other?  I've only just noticed you got a global one and then one for each package type?  I assume the global one is overridden when any of the type specific ones are set?  If so, does setting any of the type specific ones to 0 turn caching off because the notes for each omit that unlike the global one?

**Comment by pflannery** [comment]
The global options are overridden by the provider options (if present). 
Providers inherit\fallback to the global options. This only effects the duration of the cache.
There is cache map for per http client and per cli client.

I should log the cache key so it can be better understood..its in the format of something like `GET_{url}?{query}`..

To be honest my plan was to move caching to the core package request area. All providers go through that part of the core app. Each provider will have it's own map with `packagePath_packageName_packageVersion` as the key. 

When I tried this last week I found that the clean architecture chore will have to come first as the older version of the app 0.x.x is still lingering in places (moving away bit by bit). The original prototype was crazy complex to make change.
